### ISSUE TITLE
Managing stateful updates

### ISSUE THREAD

**Issue by ebenolson** [Lasagne/Lasagne#571]
This is probably obvious, but I'd never thought about it before and got bit recently. If you are using any of the updates besides `sgd`, they introduce additional state beyond the layer params. This is a consideration when trying to do checkpointing for example.

I'm not sure if any changes (`get_all_updates_values`?) make sense at the moment, but it's something to keep in mind when/if training loop code is added.

**Comment by f0k** [comment]
Well, what we'd need to do would be to scrape all update dictionary keys that are not already included in `get_all_params()`. This would be easy to add to the end of `get_all_param_values()`, with an optional `updates` keyword, except if somebody provides `**tags`.
Hmm... maybe we should promote that  for checkpointing the best way is to just get out all the values of the `updates` dictionary? This will naturally include both the trainable parameters and the optimizer state, so for this use case we don't need to distinguish the two. `get_all_updates_values()` could work for this, or simply `get_all_values(updates.values())` and `set_all_values(updates.values())`, with two additional helper functions `get_all_values()` and `set_all_values()` that'd also be used by `get_all_param_values()` and `set_all_param_values()`.

**Comment by benanne** [comment]
Sounds like this could be useful indeed! I never paid attention to this because people never used anything beyond momentum and restarting from a checkpoint with zero momentum usually didn't hurt that much. But I guess with more advanced learning rules (and things like batchnorm updating running means / variances) it starts to matter more.

**Comment by f0k** [comment]
> I never paid attention to this because people never used anything beyond momentum and restarting from a checkpoint with zero momentum usually didn't hurt that much.

You can also run into the opposite problem: If you're doing cross-validation and only reset the network weights, the momentum has still got some knowledge about data it shouldn't have (although it's probably rather meaningless when applied to a freshly initialized network).

> things like batchnorm updating running means / variances

They're part of the network parameters, so they're good.

**Comment by benanne** [comment]
Right :)
